regression-tests/run-tests.sh

```diff
@@ -129,31 +129,39 @@ if [[ "$cxx_compiler" == *"cl.exe"* ]]; then
     exec_out_dir="$expected_results_dir/msvc-2022"
     compiler_version=$(cl.exe)
 else
-    compiler_cmd="$cxx_compiler -I../../../include -std=c++20 -pthread -o "
-    compiler_version=$("$cxx_compiler" --version)
-
-    # We don't currently support Apple Clang 15 so try and switch to 14
-    if [[ "$compiler_version" == *"Apple clang version 15.0"* ]]; then
-        printf "Found Apple Clang 15, attempting to switch to Apple Clang 14"
-        cxx_compiler=$(xcodebuild -find clang++)
-        compiler_version=$("$cxx_compiler" --version)
+    # Verify the compiler command
+    which "$cxx_compiler" > /dev/null
+    if [[ $? != 0 ]]; then
+        printf "The compiler '$cxx_compiler' is not installed\n\n"
+        exit 2
     fi
 
-    if [[ "$compiler_version" == *"Apple clang version 14.0"* ]]; then
+    cpp_std=c++2b
```

Comment:
The tests should be run with C++20.
This makes it not do so, right?
We can't even have tests for C++23 (e.g., see commit 76f6d52c0ec9b9690bbac60ec829d3ccc133f68c).

Reply:
Oh, sorry, I didn't know. The confusion came from the fact that `pure2-assert-expected-not-null.cpp2` uses `std::expected` which is a C++-23 feature. I also saw PRs adding more support for `std::expected` to cppfront.
All tests currently only work for MSVC which uses `c++latest`. This seems to mean C++23 (at least for `std` implementation) given that the test with `std::expected` does compile. As I mentioned it the PR description using `c++2b` makes this test compile and work for GCC 13 as well.
As you can see from the fact that all the GitHub actions succeed, `c++2b` works for all compilers on Ubuntu 2204 for the current tests. Only GCC 10 doesn't support that option. Clang 15 also has problems with this option but this seems to be a GitHub runner issue.
Given that it is 2024 already and C++26 is WIP I think it makes a lot of sense to start using C++23 features for compilers that support them and look into compilers that support even more of them.
This PR allows to select the C++ standard used for each compiler.
It is already the case that compilation issues are committed as expected test results for certain compilers. The test scripts lists all such cases at the end. In fact, currently no compiler on POSIX platforms can compile the lowered code for all the tests. I would argue it will be very hard for a future-oriented project like cppfront to either only support older features or support newer features but not to have tests of that support.
All that being said, I believe this is @hsutter's call in the end.

Reply:
@hsutter I you are OK with compiling all regression tests with `c++2b` for compilers that support that this PR is ready for merging, I you prefer to revert to `c++20` I can do that.